### PR TITLE
Fix build with Boost 1.86.0

### LINKED ISSUE #2378: Build fails with boost 1.86
With Boost 1.86.0-2, I get the following error during build:
```
In file included from src/exprbase.h:56,
                 from src/draft.h:44,
                 from src/stats.cc:34
src/utils.h: In function ‘ledger::string ledger::sha1sum(const string&)’:
src/utils.h:602:18: error: cannot convert ‘unsigned int [5]’ to ‘unsigned char (&)[20]’
  602 |   sha.get_digest(message_digest);
      |                  ^~~~~~~~~~~~~~
      |                  |
      |                  unsigned int [5]
In file included from src/utils.h:46:
/usr/include/boost/uuid/detail/sha1.hpp:179:43: note:   initializing argument 1 of ‘void boost::uuids::detail::sha1::get_digest(unsigned char (&)[20])’
  179 | inline void sha1::get_digest(digest_type& digest)
      |                              ~~~~~~~~~~~~~^~~~~~
```

### PR DESCRIPTION
**Pull request by dbear496**
A recent update to Boost changed the type of message digests from `int[5]` to `char[20]`. This PR updates the handling of message digests to use the updated type.

Fixes #2378.

### CODE REVIEW

File: src/filters.cc [hunk 1]
```diff
@@ -238,7 +238,7 @@ void anonymize_posts::render_commodity(amount_t& amt)
 void anonymize_posts::operator()(post_t& post)
 {
 	boost::uuids::detail::sha1  sha;
-  unsigned int message_digest[5];
+  unsigned char message_digest[20];
```
Comment (afh) [inline src/filters.cc:241]:
```suggestion
#if BOOST_VERSION >= 108600
  unsigned char message_digest[20];
#else
  unsigned int message_digest[5];
#endif
```

File: src/utils.h [hunk 2]
```diff
@@ -578,14 +578,14 @@ inline int peek_next_nonws(std::istream& in) {
     *_p = '\0';                                         \
   }
 
-inline string to_hex(unsigned int * message_digest, const int len = 1)
+inline string to_hex(unsigned char * message_digest, const int len = 4)
```
Comment (afh) [inline src/utils.h:581]:
```suggestion
#if BOOST_VERSION >= 108600
inline string to_hex(unsigned char * message_digest, const int len = 4)
#else
inline string to_hex(unsigned int * message_digest, const int len = 1)
#endif
```

**Review by afh — CHANGES_REQUESTED** [review]
Thanks for addressing build issue with boost 1.86.0 and putting up this PR, @dbear496, much appreciated.

The Ledger project somewhat prides itself with maintaining compatibility with older versions of its dependencies.

Please find below a few suggestions that allow ledger to be build with boost 1.86.0 and previous versions. This should also address the CI build failures.

File: src/filters.cc (hunk 1 above)
Comment (dbear496) [reply src/filters.cc]: Changed to use `sha1sum` directly.

File: src/utils.h (hunk 2 above)
Comment (dbear496) [reply src/utils.h]: Changed to use `boost::uuids::detail::sha1::digest_type` for `message_digest`. And changed the default value of `len` to be the size of `boost::uuids::detail::sha1::digest_type` in nibbles. `len` now controls the number of hex characters in the digest string.

File: src/utils.h [hunk 3]
```diff
@@ -578,14 +578,14 @@ inline int peek_next_nonws(std::istream& in) {
...
 {
   std::ostringstream buf;
 
-  for(int i = 0; i < 5 ; i++) {
-    buf.width(8);
+  for(int i = 0; i < 20 ; i++) {
+    buf.width(2);
```
[the remark this answers is not in the document]
Comment (dbear496) [reply src/utils.h]: Changed to automatically select the right parameters based on the shape of `boost::uuids::detail::sha1::digest_type`.

File: src/utils.h [hunk 4]
```diff
@@ -598,9 +598,9 @@ inline string sha1sum(const string& str)
 
   sha.process_bytes(str.c_str(), str.length());
 
-  unsigned int message_digest[5];
+  unsigned char message_digest[20];
```
[the remark this answers is not in the document]
Comment (dbear496) [reply src/utils.h]: Changed to use `boost::uuids::detail::sha1::digest_type`.

File: src/utils.h [hunk 5]
```diff
@@ -598,9 +598,9 @@ inline string sha1sum(const string& str)
...
   sha.get_digest(message_digest);
-  return to_hex(message_digest, 5);
+  return to_hex(message_digest, 20);
```
[the remark this answers is not in the document]
Comment (dbear496) [reply src/utils.h]: Changed to use the size of `boost::uuids::detail::sha1::digest_type` in nibbles.

**Comment by dbear496** [comment]
I made it compatible with older versions of Boost by using Boost's `digest_type` instead of hard-coding the underlying type we expect Boost to use.

**Review by jwiegley — APPROVED** [review]

**Comment by jwiegley** [comment]
It looks like this PR may need to be rebased onto the current master, since I'm seeing changes here that were merged in by another PR.

**Comment by jwiegley** [comment]
@afh Does this PR look good to you now?

**Comment by afh** [comment]
I'm doing test builds with boost 1.86 and a previous version, while having a closer look at the changes this very moment :)

**Comment by afh** [comment]
While this PR builds with a previous version of boost, it unfortunately fails when building with boost 1.86.0 with:

```
src/utils.h:602:18: error: non-const lvalue reference to type 'digest_type' (aka 'unsigned char[20]') cannot bind to a value of unrelated type 'unsigned int[5]'
    sha.get_digest(message_digest);
                   ^~~~~~~~~~~~~~
  include/boost/uuid/detail/sha1.hpp:179:43: note: passing argument to parameter 'digest' here
  inline void sha1::get_digest(digest_type& digest)
```

Does this ring a bell, @dbear496?

**Comment by afh** [comment]
Oh wait, I was on the wrong git branch, that did not have this PRs changes applied. Apologies, let me try again, 😅

**Review by afh — APPROVED** [review]
Great work, @dbear496, thanks for the improvements and your contribution!! 👍 

Happy to report that this PR builds with boost 1.86.0 and previous version (I tested using 1.81.0) and all the tests passed.
